src/gui/render.cpp

```diff
@@ -725,30 +722,299 @@ static const char* to_string(const enum MonochromePalette palette)
 	}
 }
 
-bool RENDER_IsAspectRatioCorrectionEnabled()
+static AspectRatioCorrectionMode aspect_ratio_correction_mode = {};
+
+static AspectRatioCorrectionMode get_aspect_ratio_correction_mode_setting()
 {
-	return get_render_section()->Get_bool("aspect");
+	const std::string mode = get_render_section()->Get_string("aspect");
+
+	if (has_true(mode) || mode == "auto") {
+		return AspectRatioCorrectionMode::Auto;
+
+	} else if (has_false(mode) || mode == "square-pixels") {
+		return AspectRatioCorrectionMode::SquarePixels;
+
+	} else if (mode == "stretch") {
+		return AspectRatioCorrectionMode::Stretch;
+
+	} else {
+		LOG_WARNING("RENDER: Invalid 'aspect' setting '%s', using 'auto'",
+		            mode.c_str());
+		return AspectRatioCorrectionMode::Auto;
+	}
+}
+
+AspectRatioCorrectionMode RENDER_GetAspectRatioCorrectionMode()
+{
+	return aspect_ratio_correction_mode;
 }
 
 static IntegerScalingMode get_integer_scaling_mode_setting()
 {
 	const std::string mode = get_render_section()->Get_string("integer_scaling");
 
-	if (mode == "off") {
+	if (has_false(mode)) {
 		return IntegerScalingMode::Off;
+
 	} else if (mode == "auto") {
 		return IntegerScalingMode::Auto;
+
 	} else if (mode == "horizontal") {
 		return IntegerScalingMode::Horizontal;
+
 	} else if (mode == "vertical") {
 		return IntegerScalingMode::Vertical;
+
 	} else {
 		LOG_WARNING("RENDER: Invalid 'integer_scaling' setting: '%s', using 'auto'",
 		            mode.c_str());
 		return IntegerScalingMode::Auto;
 	}
 }
 
+static void set_default_viewport_setting()
+{
+	const auto string_prop = get_render_section()->GetStringProp(
+	        "viewport");
+	string_prop->SetValue("fit");
+}
+
+static void log_invalid_viewport_setting_warning(
+        const std::string& pref,
+        const std::optional<const std::string> extra_info = {})
+{
+	LOG_WARNING("DISPLAY: Invalid 'viewport' setting: '%s'"
+	            "%s%s, using 'fit'",
+	            pref.c_str(),
+	            (extra_info ? ". " : ""),
+	            (extra_info ? extra_info->c_str() : ""));
+}
+
+std::optional<std::pair<int, int>> parse_int_dimensions(const std::string_view s)
+{
+	const auto parts = split(s, "x");
+	if (parts.size() == 2) {
+
+		const auto w = parse_int(parts[0]);
+		const auto h = parse_int(parts[1]);
+		if (w && h) {
+			return {{*w, *h}};
+		}
+	}
+	return {};
+}
+
+static std::optional<ViewportSettings> parse_fit_viewport_modes(const std::string& pref)
+{
+	if (pref == "fit") {
+		ViewportSettings viewport = {};
+		viewport.mode             = ViewportMode::Fit;
+		return viewport;
+
+	} else if (const auto width_and_height = parse_int_dimensions(pref)) {
```

Comment:
the 'if' body looks a bit strange without knowing that `parse_int_dimensions()` returns an optional.

For `<optional>` return values, I like to prefix the variables with `maybe_` because the variable might not exist, and then strip off the `maybe_` once we've confirmed it's available:

(this is a good use of the C++17's `if (initializer; conditional)` syntax where we can test the optional in the condition):

```c++
if (const auto maybe_w_and_h = parse_dims(pref); maybe_w_and_h) {
    const auto [w, h] = *maybe_w_and_h;
    ...
}
```

Reply:
Yeah, I guess I can, although it's a bit wordy. In `programs_mixer.cpp` I've just done something like this in many places:

```ini
if (const auto v = parse_volume(parts[0]); v) {
  const SetVolume cmd = {AudioFrame(*v, *v)};
```

Might seem strange a bit first, but people quickly get used to it that the `parse_*` functions all return options.

Maybe I'll do the "maybe" here 😄 , because there's only a few `parse` invocations. If there were many, I'd just say people need to learn how the `parse` functions work.

Reply:
That's a lot of maybes!

Also no worries - having the test `(... ; v) {` is the self-documenting part -> so that works too!

Reply:
Adding `maybe_*` prefix in `parse_relative_viewport_modes()` because it clarifies things there. Leave the rest as-is when I'm checking for the optional value being set in `if` statements.